brontide: retry enabling channels

yyforyongyu opened the pull request:
This PR fixes a bug that are observed in itest failures,
- after reconnection, a node may fail to enable its channel and thus won't announce it to the network. This happens when the startup of the link is falling behind, in specific, the `reestablished` has not been marked in `htlcManager` while the peer is requesting the enabling. This could happen when `chan-enable-timeout` is set to a small value, while the link busy syncing the states. We fix it by retrying the request after 5 seconds, in the hope that the link is now ready.

GeorgeTsagk commented on peer/brontide.go:
```diff
@@ -2105,52 +2115,94 @@ func (p *Brontide) queueHandler() {
-				} else {
-					lazyMsgs.PushBack(msg)
-				}
+				// We know the elem must from priority queue.
```
nit: comment typo / missing words?

GeorgeTsagk commented on peer/brontide.go:
```diff
@@ -2105,52 +2115,94 @@ func (p *Brontide) queueHandler() {
+				"queue: %v", elem)
+		}
+
+		p.log.Tracef("Sending laze msg to sendQueue: %s",
```
nit: typo "laze" -> "lazy"

GeorgeTsagk commented on peer/brontide.go:
```diff
@@ -2105,52 +2115,94 @@ func (p *Brontide) queueHandler() {
+		// send the message to `sendQueue`.
+		select {
+		case p.sendQueue <- front:
+			// We know the elem must from lazy queue.
```
nit: same, sentence typo / missing words?

GeorgeTsagk commented on peer/brontide.go:
```diff
@@ -2105,52 +2115,94 @@ func (p *Brontide) queueHandler() {
+				// messages again if the queue is not empty,
+				// and we won't accept new messages until it's
+				// empty.
+				continue
```
So for a given priorityMsg queue you first drain it before new messages get accepted, even priority messages?

yyforyongyu replied:
yeah i want the old priority messages to get processed first.

GeorgeTsagk commented on peer/brontide.go:
```diff
@@ -2105,52 +2115,94 @@ func (p *Brontide) queueHandler() {
+
+		// If there are no messages in the lazy queue, we'll block
+		// until a new message is received.
+		if elem == nil {
```
Why are we first checking if `elem` is `nil` in order to block for any message, and then handle the lazy message outside of an `if` block?

Unless there is an algorithmic reason not to, I believe it would be more consistent/readable to first check if `elem != nil` and proceed with the lazy sending as written below, but this time inside of the `if` block. At the end there could be the generic blocking code that waits for any message as the loop's last resort, outside of any `if` section.

Just personal opinion.

GeorgeTsagk replied:
So the flow would be:

1. Do a priority-send, if priority queue not empty
2. Do a lazy-send, if lazy queue not empty
3. Wait for any message, if both queues are empty

GeorgeTsagk replied:
Since lazy-send also waits for new messages there wouldn't be any difference in the overall behavior of the loop.

yyforyongyu replied:
so you mean changing the `if elem == nil` to `if elem != nil`?

> Why are we first checking if elem is nil in order to block for any message, and then handle the lazy message outside of an if block?

This is actually more for readability. The code space is less wrapped this way since we need to check more using `if elem != nil`.

GeorgeTsagk replied:
> so you mean changing the if elem == nil to if elem != nil?

yes, and executing the lazy-send inside the if block

> This is actually more for readability. The code space is less wrapped this way since we need to check more using `if elem != nil`.

Ok

GeorgeTsagk commented on peer/brontide.go:
```diff
@@ -2538,13 +2590,59 @@ func (p *Brontide) reenableActiveChannels() {
+				chanPoint, false,
+			)
+			if err == nil {
+				continue
```
So we wait for `retryEnableInterval` only once before exiting and giving up on enabling the channel?

Crypt-iQ replied:
5 seconds is a lot, but it should retry

yyforyongyu replied:
> giving up on enabling the channel?

Yeah we'd give up. The only reason we give it a second chance is because the startup order of our subsystems may cause this "flake", and we wait and try. 5s is a lot, only added for itest and I think 1s should be enough.

GeorgeTsagk reviewed (commented):
Clean work! :100: just a few comments :)

GeorgeTsagk commented on peer/brontide.go:
```diff
@@ -2538,13 +2590,59 @@ func (p *Brontide) reenableActiveChannels() {
 	// channel is already active, the update won't be sent.
 	for _, chanPoint := range activePublicChans {
 		err := p.cfg.ChanStatusMgr.RequestEnable(chanPoint, false)
```
Can `ChanStatusMgr.RequestEnable(chanPoint, false)` be called concurrently from multiple go routines?

If yes, then a re-work could be considered where the extra attempts to enable a channel are done in a go routine, possibly attempting multiple times within a wait duration, with attempts occurring every `retryEnableInterval`. This way the attempts would be more persistent. Of course this would be more cpu intensive, only to be considered as a work-around if a single `retryEnableInterval` is not long enough to wait for the link activation line.

yyforyongyu replied:
`reenableActiveChannels` will only be called on start once so nope.

Crypt-iQ commented on peer/brontide.go:
```diff
@@ -2111,52 +2111,94 @@ func (p *Brontide) queueHandler() {
-			front := elem.Value.(outgoingMsg)
+			front, ok := elem.Value.(outgoingMsg)
+			if !ok {
+				p.log.Errorf("Unknown message type found in "+
```
return err?

GeorgeTsagk replied:
+1, should we proceed with sending if the msg type is unknown?

Crypt-iQ commented on peer/brontide.go:
```diff
@@ -2538,13 +2544,59 @@ func (p *Brontide) reenableActiveChannels() {
+		//
+		// TODO(yy): make the link signaling back its current state to
+		// properly handle this issue.
+		case errors.Is(err, netann.ErrEnableInactiveChan):
```
the link calls `l.cfg.NotifyActiveChannel(*l.ChannelPoint())` when it's re-established which gives an ActiveChannelEvent via `SubscribeChannelEvents` but it's possible the notification will be missed. You could export `isReestablished` and make it a `ChannelLink` interface function and then check that

Crypt-iQ replied:
I think you could:
- `SubscribeChannelEvents`
- call `RequestEnable`
- if `RequestEnable` fails, we know the link isn't active yet, so we continue listening for `ActiveChannelEvent`
- if `RequestEnable` succeeds, stop listening for the event and continue

yyforyongyu replied:
yeah I think the two systems are not on the same page. Somehow the `activePublicChans` here is not considered active by htlcswitch. 

> You could export isReestablished and make it a ChannelLink interface function and then check that

if the `htlcManager` hasn't marked it then this won't make a difference?

I think my question is, can any of the channels inside `activePublicChans` fail somehow? Maybe `RequestEnable` should block for some time if it sees the link is not reestablished...

Crypt-iQ replied:
> I think my question is, can any of the channels inside activePublicChans fail somehow?

What do you mean here?

` Somehow the activePublicChans here is not considered active by htlcswitch.`

Because the reestablished atomic var isn't set?

I think the steps I wrote above should resolve this... maybe if `RequestEnable` fails for reasons other than inactive link, you retry, and otherwise wait for the active channel event (which is sent after the reestablish var in the link is set

Crypt-iQ reviewed (commented):
Is the second commit (peer changes) fixing a bug? If not, I think we should leave the code as-is. Anything in the queues will eventually go through since it's a 50-50 chance whether we add a new message or process an old one. Plus I'm not a fan of changing that code

yyforyongyu commented:
> Is the second commit (peer changes) fixing a bug? If not, I think we should leave the code as-is. Anything in the queues will eventually go through since it's a 50-50 chance whether we add a new message or process an old one. Plus I'm not a fan of changing that code

hmm curious why not a fan? because it's too fundamental? I did see several itest failures go away with it, but still need to observe more and see if it really fixes anything.

Crypt-iQ commented:
> hmm curious why not a fan? because it's too fundamental? I did see several itest failures go away with it, but still need to observe more and see if it really fixes anything.

I think your patch is correct, but I think it's not something we should change unless we have to (e.g. there's a bug or some clear win). Or maybe it should be a separate PR (which I know is burdensome) so review can just be on those changes

yyforyongyu commented:
> I think your patch is correct, but I think it's not something we should change unless we have to (e.g. there's a bug or some clear win). Or maybe it should be a separate PR (which I know is burdensome) so review can just be on those changes

Cool yeah make sense, will take out that commit. If needed will create another PR.

yyforyongyu commented:
Updated the PR to include the retry fix only. Also changed how the retry logic to apply @Crypt-iQ 's solution as it's much more elegant than the previous sleep and retry.

Crypt-iQ commented on htlcswitch/link.go:
```diff
@@ -969,6 +969,9 @@ func (l *channelLink) loadAndRemove() error {
 func (l *channelLink) htlcManager() {
 	defer func() {
+		// Send an inactive notification when the link exits.
+		l.cfg.NotifyInactiveChannel(*l.ChannelPoint())
```
Ah, I see the issue with my original suggestion - the link can fail during reestablish and then the code in `peer/brontide.go` waits for an event that will never come. Instead of breaking the current flow that `NotifyInactiveChannel` is always called _with_ a `NotifyActiveChannel`, maybe this should be a new notification `NotifyInactiveLink`.

yyforyongyu replied:
cool yeah always prefer being explicit, added the suggested event now.

Crypt-iQ commented on peer/brontide.go:
```diff
@@ -2506,19 +2507,61 @@ func (p *Brontide) reenableActiveChannels() {
+	//
+	// NOTE: ChannelNotifier must be started before subscribing events
+	// otherwise we'd panic here.
+	sub, err := p.cfg.ChannelNotifier.SubscribeChannelEvents()
```
I think this is a bit tricky, I'm pretty sure it's possible that:
- the link could have called `fail(...)`
- sent the `NotifyInactiveChannel` notification
- and _still_ be in the `activeChannels` map since the `handleLinkFailure` function may not have run yet in `peer/brontide.go`

This means that we would have missed the `NotifyInactiveChannel` notification and wait indefinitely for an event that won't come. Not entirely sure what to do here

yyforyongyu replied:
I'm thinking of maybe two possible solutions,
1. make the `onChannelFailure` blocking on `handleLinkFailure`, which means we need the method to signal back when it's finished. This way the `activeChannels` would be accurate.
2. make the subscription even earlier, maybe inside `addLink` so we won't miss any events.

I think an extension of solution 2 is to expand the usage of `ChannelNotifier` inside `peer` and we can craft actions based on the event, sort like the only communication channel between `link` and `peer`. Ofc this is just a thought, would require dramatic change I think.

Crypt-iQ replied:
I think having two subsystems communicate with each other in the current way sucks because I've found and also introduced hard-to-debug deadlocks. So I don't like 1 unless we just change how we do it and NOT use go chans. Option 2 seems like a better idea because then we won't be able to miss any events, but also not perfect because we have to keep track of this notifier for each link.

If we go with option 2, there is one place in the peer that already subscribes to channel events (https://github.com/lightningnetwork/lnd/blob/6f944b13c8f7d139f6592f40a22e47deb94028ff/peer/brontide.go#L1298-L1310), so it may be worth it to have a dedicated `peer/brontide.go` goroutine to listen and pass the events to the other relevant peer goroutines.

yyforyongyu replied:
> so it may be worth it to have a dedicated peer/brontide.go goroutine to listen and pass the events to the other relevant peer goroutines.

So I tried to implement this as I like that a) we make early subscriptions to avoid missing events and b) no more one link one subscription, now all links share the same subscription.

But the issues are,
1) since each event is sent only once, what if the `ActiveLinkEvent` is wanted at both places?
2) if we decide to make the subscription early in the `(p *Brontide) Start()`, how do we deal with the old updates?

In the current approach, the subscription is canceled quickly, so I guess that's a pro. But it's prone to miss events, and that's a big con.

Crypt-iQ replied:
For 1. if we had a dedicated goroutine in `peer/brontide.go`, we could duplicate the events and send it to the other goroutines that needed the event. Or we could just have multiple subscriptions per link - this option seems a bit easier.

For 2. not sure why old updates matter? Think we could ignore them?

Crypt-iQ replied:
what is an old update here?

yyforyongyu replied:
a link can be activated and deactivated, then there would be a pair of events? By the time you want to check for `waitUntilLinkActive`, or any other places, how do you know if you are not receiving an old event?

Crypt-iQ replied:
I don't think what you're describing can happen:

- `NotifyActiveLink` in `link.go` calls `SendUpdate`
  - The call to `SendUpdate` will block until the underlying `ntfnServer` is able to pull the update from the `s.updates` chan in `subscribe/subscribe.go`
  - When the update is pulled, the update is sent out to every client in the `s.clients` map.
- For a new peer object to be up that is also interested in this `ActiveLinkEvent`, the links must all be stopped beforehand. This is in `peerTerminationWatcher` in `server.go`.
  - If this new peer object were to receive an old event, it would need to already be in the `s.clients` map _at the same time_ that the update is pulled off the `s.updates` chan in `subscribe/subscribe.go`.
  - This isn't possible since the peer object would need to subscribe _before or at the same time_ that the `ActiveLinkEvent` went out. But the prior links must be stopped for this peer object to be started which is a contradiction

yyforyongyu commented:
I think this is missing some unit tests. Will create a follow up PR to interface `ChannelNotifier` and add tests.

lightninglabs-deploy commented:
@yyforyongyu, remember to re-request review from reviewers when ready

yyforyongyu commented:
I think the current approach can be resource hungry as it's `O(N^2)`. If we have 100 channels, each has one update, then this would create 100 subscriptions, and each subscription gets 100 events. This can be mitigated by adding a new subscription method where we only subscribe for a specific channel link. Moreover, compared to the old way used in `waitUntilLinkActive`, the subscription now needs to be alive for the whole lifecycle of the peer.

I think a better way to handle this is to make a single subscription early in `peer.Start()` so we don't miss events. And use it to handle the re-enabling channels only. Once re-enabling is done, this subscription can be canceled.

Crypt-iQ commented on peer/brontide.go:
```diff
@@ -2309,7 +2320,9 @@ func (p *Brontide) genDeliveryScript() ([]byte, error) {
-func (p *Brontide) channelManager() {
+func (p *Brontide) channelManager(client *subscribe.Client) {
+	// When exit, cancel the channel event subscription.
+	defer client.Cancel()
```
Since the `reenableTimeout` only fires once, this `Cancel` should be called in that same block when the `reenableTimeout` is set to `nil`. If we call `Cancel` here, then this will cause updates to queue up in the internal overflow queue, possibly leading to DoS. This is because we may be done receiving channel events, but the p2p connection is still up and therefore the `channelManager` goroutine is still up.

Crypt-iQ commented on peer/brontide.go:
```diff
@@ -592,6 +593,16 @@ func (p *Brontide) Start() error {
+	//
+	// NOTE: ChannelNotifier must be started before subscribing events
+	// otherwise we'd panic here.
+	sub, err := p.cfg.ChannelNotifier.SubscribeChannelEvents()
```
I think the fix will work as intended, but the reenableTimeout is 19 minutes by default. I'm wondering if it's possible that we receive many channel events before then that stack up in the notifier's overflow queue and cause a DoS. Maybe this could happen if many peers have flapping connections and multiple links each.

yyforyongyu replied:
Maybe we could do a check on `reenableTimeout`, if it's less than, say 2 min, then we make this subscription, otherwise we won't bother retrying enabling the channel again?

Crypt-iQ reviewed (commented):
Left two comments, I think the approach is right just need to make sure there isn't any side effect

GeorgeTsagk commented on peer/brontide.go:
```diff
@@ -2506,48 +2539,62 @@ out:
+		// If the channel is reported as inactive, we will give it
+		// another chance. When handling the request, ChanStatusManager
+		// will check whether the link is active or not. One of the
+		// conditions is whether the link as been marked as
```
nit: typo "the link as been marked" -> "the link has been marked"

yyforyongyu replied:
good catch, fixed!

GeorgeTsagk reviewed (approved):
LGTM 👍 

Only got a few comments for nits, the retrying design seems good :)

yyforyongyu commented on peer/brontide.go:
```diff
@@ -2626,6 +2673,136 @@ func (p *Brontide) fetchActiveChanCloser(chanID lnwire.ChannelID) (
 }
 
+// filterChannelsToEnable filters a list of channels to be enabled upon start.
+// The filtered channels are active channels that's neither privite nor
```
fixed!

lightninglabs-deploy commented:
@crypt-iq: review reminder

lightninglabs-deploy commented:
@crypt-iq: review reminder

lightninglabs-deploy commented:
@crypt-iq: review reminder

lightninglabs-deploy commented:
@crypt-iq: review reminder

Crypt-iQ reviewed (approved):
this seems like an unrelated flake? https://github.com/lightningnetwork/lnd/actions/runs/3710090005/jobs/6289289320

yyforyongyu commented:
> this seems like an unrelated flake? https://github.com/lightningnetwork/lnd/actions/runs/3710090005/jobs/6289289320

have seen this in other builds too and will investigate the test `TestHistoricalConfDetailsTxIndex`, but I don't it's related.

yyforyongyu commented on peer/brontide.go:
```diff
@@ -2628,6 +2601,44 @@ func (p *Brontide) fetchActiveChanCloser(chanID lnwire.ChannelID) (
+
+		dbChan := lnChan.State()
+		isPublic := dbChan.ChannelFlags&lnwire.FFAnnounceChannel != 0
+		if !isPublic || dbChan.IsPending {
```
Dope, will create an issue to track it!
